### ISSUE TITLE
Future epic: add typed model/provider registry and shared inference DI

### ISSUE THREAD

**Issue by dayvidpham** [ananyasingh8/llm-rudeness-preferences#1]
## Problem

Model selection and runtime construction are currently experiment-specific:

- Quadratic voting pins and constructs a local Transformers model directly.
- Bail augmentation constructs an OpenRouter client and passes a raw model slug.
- Future emotion probes will require the actual local PyTorch model and tokenizer for activation access.

This makes it easy for experiments to use inconsistent identities or invalid model/provider combinations. Failures can occur deep inside an experiment instead of once at a validated construction boundary.

## User Impact

A small shared registry should let all three experiments:

- select model and provider identities consistently;
- swap reviewed quantizations of the same model without changing experiment logic;
- inject text generation without importing provider SDKs;
- preserve local Transformers activation access for probe work;
- reject unsupported routes before an experiment starts;
- keep model identity separate from artifact revision, provider, credentials, and per-run generation settings.

## Current State

- The workstreams are bail behavior, emotion probes, and quadratic voting (`README.md`).
- QV pins `google/gemma-4-E2B-it-qat-q4_0-unquantized` at revision `6befbaca7398925921802abd1f277b495b78b738` and owns download/loading/generation directly (`quadratic_voting/main.py`).
- Bail stores raw OpenRouter URL/model strings and constructs its client inside the pipeline (`bail/config.py`, `bail/src/augment_bailbench.py`).
- Bail credentials can fall back to `secrets_local.py`; the shared provider boundary should accept `OPENROUTER_API_KEY` only.
- Emotion probes are not implemented yet, but they must require an activation-capable local runtime rather than a generic remote generator.

## Proposed Minimal API

Add a closed `llm_runtime/` package based on enums, frozen dataclasses, mappings, and structural protocols. Do not add a provider/model class hierarchy, plugin framework, service locator, or arbitrary model-slug escape hatch.

### Shared Types

```python
class ModelId(StrEnum):
    GEMMA_4_E2B_IT = "gemma-4-e2b-it"
    DOLPHIN_MISTRAL_24B_VENICE = "dolphin-mistral-24b-venice"


class ProviderId(StrEnum):
    LOCAL = "local"
    OPENROUTER = "openrouter"


class QuantizationId(StrEnum):
    BF16 = "bf16"
    W4A16_COMPRESSED_TENSORS = "w4a16-compressed-tensors"


class RuntimeId(StrEnum):
    TRANSFORMERS = "transformers"
    OPENAI_COMPATIBLE_HTTP = "openai-compatible-http"


class Capability(StrEnum):
    TEXT_GENERATION = "text-generation"
    LOCAL_ACTIVATIONS = "local-activations"


class MessageRole(StrEnum):
    SYSTEM = "system"
    USER = "user"
    ASSISTANT = "assistant"


@dataclass(frozen=True, slots=True)
class ChatMessage:
    role: MessageRole
    content: str


@dataclass(frozen=True, slots=True)
class GenerationSettings:
    max_new_tokens: int
    temperature: float = 0.0


class TextGenerator(Protocol):
    @property
    def model_id(self) -> ModelId: ...

    @property
    def provider_id(self) -> ProviderId: ...

    @property
    def quantization_id(self) -> QuantizationId: ...

    def generate(
        self,
        messages: Sequence[ChatMessage],
        settings: GenerationSettings,
    ) -> str: ...
```

`GenerationSettings` validates positive token counts and finite, non-negative temperature at construction. It remains per-run state and is not stored in the registry.

### Closed Registry

```python
type TransformersModelId = Literal[ModelId.GEMMA_4_E2B_IT]
type OpenRouterModelId = Literal[ModelId.DOLPHIN_MISTRAL_24B_VENICE]
type TransformersQuantizationId = Literal[
    QuantizationId.BF16,
    QuantizationId.W4A16_COMPRESSED_TENSORS,
]


@dataclass(frozen=True, slots=True)
class HuggingFaceArtifact:
    repository: str
    revision: str


@dataclass(frozen=True, slots=True)
class LocalTransformersRoute:
    model_id: TransformersModelId
    quantization_id: TransformersQuantizationId
    artifact: HuggingFaceArtifact
    context_window: int
    provider_id: Literal[ProviderId.LOCAL] = ProviderId.LOCAL
    runtime_id: Literal[RuntimeId.TRANSFORMERS] = RuntimeId.TRANSFORMERS
    capabilities: frozenset[Capability] = frozenset({
        Capability.TEXT_GENERATION,
        Capability.LOCAL_ACTIVATIONS,
    })


@dataclass(frozen=True, slots=True)
class OpenRouterRoute:
    model_id: OpenRouterModelId
    quantization_id: None = None
    model_slug: str
    provider_id: Literal[ProviderId.OPENROUTER] = ProviderId.OPENROUTER
    runtime_id: Literal[RuntimeId.OPENAI_COMPATIBLE_HTTP] = (
        RuntimeId.OPENAI_COMPATIBLE_HTTP
    )
    capabilities: frozenset[Capability] = frozenset({
        Capability.TEXT_GENERATION,
    })


type ModelRoute = LocalTransformersRoute | OpenRouterRoute


def resolve_route(
    model_id: ModelId,
    provider_id: ProviderId,
    quantization_id: QuantizationId | None,
    *,
    required: AbstractSet[Capability] = frozenset(),
) -> ModelRoute: ...
```

Initially enable only the two routes already evidenced in the repository. Keep W4A16 as the first candidate route until its compatibility gate passes:

| Model | Provider | Quantization | Runtime | Status | Capabilities |
|---|---|---|---|---|---|
| `GEMMA_4_E2B_IT` | `LOCAL` | `BF16` | `TRANSFORMERS` | enabled | text generation, local activations |
| `GEMMA_4_E2B_IT` | `LOCAL` | `W4A16_COMPRESSED_TENSORS` | `TRANSFORMERS` | candidate pending compatibility test | text generation; activation coverage must be tested |
| `DOLPHIN_MISTRAL_24B_VENICE` | `OPENROUTER` | none/undisclosed | `OPENAI_COMPATIBLE_HTTP` | enabled | text generation |

The registry key is `(ModelId, ProviderId, QuantizationId | None)`. Model identity must not encode quantization, and quantization must not be inferred from a repository name. Each local quantization route owns its exact artifact revision and loading metadata.

Known direct factory calls should use `Literal`-constrained model and quantization aliases. Dynamic CLI/config values should pass through `resolve_route()` exactly once. Unsupported combinations should raise an actionable `ModelRouteError` containing the requested route, quantization, missing capability, supported alternatives, and corrective action.

The current production route is `GEMMA_4_E2B_IT + LOCAL + BF16`, backed by `google/gemma-4-E2B-it-qat-q4_0-unquantized`. Treat QAT training target and runtime weight representation as different facts: this artifact was trained for Q4_0 conversion but is stored and executed as BF16.

The first additional candidate is `GEMMA_4_E2B_IT + LOCAL + W4A16_COMPRESSED_TENSORS`, backed by Google's official `google/gemma-4-E2B-it-qat-w4a16-ct`. Add that production route only after an integration test proves the selected Transformers/runtime revision can load and generate from it. The registry design must allow the variant even if compatibility validation defers enabling it.

### Provider Boundaries

```python
class LocalActivationRuntime(TextGenerator, Protocol):
    @property
    def provider_id(self) -> Literal[ProviderId.LOCAL]: ...

    @property
    def model(self) -> torch.nn.Module: ...

    @property
    def tokenizer(self) -> PreTrainedTokenizerBase: ...


def create_transformers_runtime(
    model_id: TransformersModelId,
    quantization_id: TransformersQuantizationId,
    *,
    cache_dir: Path,
    device: Device = Device.AUTO,
) -> LocalActivationRuntime: ...


@dataclass(frozen=True, slots=True, repr=False)
class OpenRouterCredentials:
    api_key: str = field(repr=False)


def openrouter_credentials_from_env(
    environ: Mapping[str, str] | None = None,
) -> OpenRouterCredentials: ...


def create_openrouter_generator(
    model_id: OpenRouterModelId,
    *,
    credentials: OpenRouterCredentials,
    http_client: httpx.Client | None = None,
) -> OpenRouterGenerator: ...
```

The local runtime keeps the real PyTorch model and tokenizer for hooks, hidden states, NNsight, and future probes. It dispatches loading from the resolved quantization route rather than branching on free-form strings. Do not invent a remote activation API.

Use `httpx` for OpenRouter's `/api/v1/chat/completions` endpoint. Accept an injected client for `httpx.MockTransport` tests. Only `OPENROUTER_API_KEY` is accepted; credentials must never appear in representations, logs, exceptions, persisted configuration, or request diagnostics.

## Dependency Direction

```text
experiment logic -> TextGenerator / LocalActivationRuntime protocols
composition roots -> provider factories
provider adapters -> registry + shared value types
registry/adapters -X-> experiment packages
```

## Minimal Implementation Plan

1. Add `llm_runtime/{__init__,types,registry,transformers,openrouter}.py` with focused tests and a quantization axis in every local route key.
2. Move the existing Gemma artifact identity, revision, BF16 quantization metadata, cache/download logic, device checks, and loader from QV into the shared Transformers adapter.
3. Refactor QV conversation logic to accept `TextGenerator`; preserve deterministic generation, context checks, and CLI defaults.
4. Refactor bail's real path to accept `TextGenerator`; replace raw provider/model constants with enums and typed settings.
5. Add `bail/__init__.py` and use root package execution so bail can import the shared package.
6. Declare and lock bail's existing data dependencies (`pandas` and a Parquet engine). Do not add the OpenAI SDK if `httpx` is used.
7. Validate the official Gemma W4A16 Compressed Tensors artifact with the selected runtime. Register it only if load and generation tests pass; otherwise record the compatibility blocker while preserving the typed extension point.
8. Update root, bail, and QV documentation with registered routes, quantization differences, capability limits, credential setup, and the process for adding a model or quantization.
9. Remove superseded string constants and direct constructors without compatibility aliases.

### Bail behavior

Inject `TextGenerator` into augmentation logic. Keep retry, resume, concurrency, and tagged-response parsing in the bail pipeline.

### Quadratic voting

Inject `TextGenerator` into conversation logic. Provider- and quantization-specific loading remains in the composition root. A CLI quantization choice is parsed into `QuantizationId` once, then resolved through the registry.

### Emotion probes

Require `LocalActivationRuntime`. Passing an OpenRouter generator should be a mypy error; dynamic selection requesting `LOCAL_ACTIVATIONS` should fail in `resolve_route()` before execution.

## Validation

- Registry tests cover every enum route and capability declaration.
- Registry tests cover valid and invalid model/provider/quantization triples and prove that two quantizations can share one `ModelId` without changing experiment APIs.
- Mypy fixtures verify invalid known factory calls are rejected.
- Runtime tests verify invalid dynamic combinations fail at `resolve_route()`.
- Transformers tests call the public factory while replacing only Hub/model-loader boundaries.
- Quantization integration tests assert each enabled variant selects its exact repository, revision, loader configuration, dtype metadata, and capability set.
- Preserve the opt-in Hugging Face metadata test and CUDA BF16 smoke test.
- OpenRouter tests use `httpx.MockTransport` to verify endpoint, slug, message serialization, settings, parsing, timeouts, malformed responses, and credential redaction.
- Bail integration tests cover message construction through the shared adapter to tagged-response parsing without network access.
- QV integration tests cover conversation state through an injected generator without loading weights.
- Activation tests verify the local factory exposes the actual `torch.nn.Module` and tokenizer.

```console
uv run python -m unittest discover -v
uv run ruff format --check .
uv run ruff check .
uv run mypy llm_runtime bail quadratic_voting
```

## Acceptance Criteria

- **Given** either registered model **when** experiment code selects it **then** it uses `ModelId` and `ProviderId` instead of raw strings **and should not** bypass the registry.
- **Given** a local model selection **when** experiment code chooses its weights **then** it uses `QuantizationId` and resolves an exact artifact revision **and should not** embed quantization in an arbitrary model slug.
- **Given** two registered quantizations for one model **when** the composition root swaps between them **then** experiment logic receives the same typed runtime protocol **and should not** branch on repository names or provider-specific strings.
- **Given** an unsupported model/provider/quantization triple **when** it is resolved **then** construction fails with supported quantizations and corrective action **and should not** silently fall back to a different precision.
- **Given** Gemma with the local provider **when** its factory is constructed **then** it resolves the existing pinned artifact and preserves activation access **and should not** make a remote request.
- **Given** Dolphin with OpenRouter **when** its generator is constructed **then** it sends the registered slug and typed settings **and should not** expose an activation interface.
- **Given** an unsupported route or capability **when** it is resolved **then** construction fails once with the route, reason, alternatives, and fix **and should not** fail later inside an experiment.
- **Given** `OPENROUTER_API_KEY` is absent or blank **when** credentials are loaded **then** construction fails before HTTP work with setup instructions **and should not** attempt a request.
- **Given** OpenRouter fails or returns malformed data **when** generation is attempted **then** an actionable error identifies remediation without authorization data **and should not** leak credentials.
- **Given** bail, QV, or another behavioral experiment **when** tested **then** a fake or real `TextGenerator` can be injected without importing provider implementations.
- **Given** emotion-probe code **when** type-checked **then** it requires `LocalActivationRuntime` **and should not** accept a remote-only generator.
- **Given** generation settings change per run **when** a request is made **then** they remain separate from model identity, artifact metadata, provider configuration, and credentials.
- **Given** migration is complete **when** quality gates run **then** they pass without model downloads or live OpenRouter requests.

## Non-Goals

- Implementing bail classification, emotion probes, or QV mechanics.
- A universal inference framework, plugin system, service locator, or model class hierarchy.
- Arbitrary user-provided model slugs.
- Arbitrary quantization names, runtime quantization recipes, or on-the-fly quantization parameters outside reviewed registry entries.
- Streaming, batching, multimodal input, embeddings, log probabilities, tool calls, or cost accounting until an experiment needs them.
- Abstracting activation hook points or pretending remote APIs expose activations.
- Adding vLLM, llama.cpp, NNsight, LiteLLM, or more providers in this epic.
- Backward-compatibility aliases for current constants or constructors.

## Risks And Decisions

- CLI/environment strings cannot be checked statically. Validate them exactly once through `resolve_route()`.
- A closed registry requires code and tests to add a model; that friction protects experimental consistency.
- A quantization enum identifies an execution artifact variant, not its training method. QAT targets, storage format, weight dtype, activation dtype, and artifact revision must remain explicit route metadata.
- OpenRouter may not disclose or guarantee an exact quantization. Its routes use `quantization_id=None` unless OpenRouter provides an enforceable contract; do not guess from marketing names.
- `TextGenerator` is intentionally the smallest common text interface. Activation access remains a separate protocol.
- OpenRouter identifies a gateway/model slug, not a reproducibly pinned weight revision. Remote runs must not be treated as equivalent to the pinned local artifact.
- Initially register only identities already used by the repository. Do not add placeholders.
- Initially enable only quantization routes that pass exact load/generation validation. The W4A16 route is the first compatibility candidate.
- Do not add a Gemma/OpenRouter route until its exact slug and routing policy are selected.
- Defer OpenRouter upstream-provider routing constraints until an experiment requires them; document that limitation in run evidence.
